internal/v1/handler_oscap.go

```diff
@@ -181,150 +182,348 @@ func (h *Handlers) GetOscapCustomizations(ctx echo.Context, distribution Distrib
 	if err == compliance.ErrorTailoringNotFound {
-		return nil, err
-	} else if err != nil {
-		return nil, echo.NewHTTPError(http.StatusInternalServerError, err)
+		return nil, nil, compliance.ErrorTailoringNotFound
+	}
+	if err != nil {
+		return nil, nil, echo.NewHTTPError(http.StatusInternalServerError, err)
+	}
+
+	var allErrors []BlueprintLintItem
+	var allWarnings []BlueprintLintItem
+
+	// Collect errors and warnings from all lint functions (they now return instead of mutating)
+	if errs, warns := lintPackages(policyBP, prevCust, bpBody, fixup); len(errs) > 0 || len(warns) > 0 {
+		allErrors = append(allErrors, errs...)
+		allWarnings = append(allWarnings, warns...)
+	}
+	if errs, warns := lintFilesystems(policyBP, prevCust, bpBody, fixup); len(errs) > 0 || len(warns) > 0 {
+		allErrors = append(allErrors, errs...)
+		allWarnings = append(allWarnings, warns...)
+	}
+	if errs, warns := lintServices(policyBP, prevCust, bpBody, fixup); len(errs) > 0 || len(warns) > 0 {
+		allErrors = append(allErrors, errs...)
+		allWarnings = append(allWarnings, warns...)
 	}
+	if errs, warns := lintKernel(policyBP, prevCust, bpBody, fixup); len(errs) > 0 || len(warns) > 0 {
+		allErrors = append(allErrors, errs...)
+		allWarnings = append(allWarnings, warns...)
+	}
+	if errs, warns := lintFIPS(policyBP, prevCust, bpBody, fixup); len(errs) > 0 || len(warns) > 0 {
+		allErrors = append(allErrors, errs...)
+		allWarnings = append(allWarnings, warns...)
+	}
+
+	return allErrors, allWarnings, nil
+}
+
+// lintPackages validates package compliance by performing two comparisons:
+// 1. ERRORS: Packages required by current policy but missing from current blueprint
+// 2. WARNINGS: Packages that were required by policy in snapshot but no longer required
+func lintPackages(policyBP *blueprint.Blueprint, prevCust, currentCust *Customizations, fixup bool) ([]BlueprintLintItem, []BlueprintLintItem) {
+	var errs []BlueprintLintItem
+	var warns []BlueprintLintItem
 
-	// make sure all packages are present, all partitions, all enabled/disabled services, all kernel args
-	for _, pkg := range bp.GetPackagesEx(false) {
-		if cust.Packages == nil || !slices.Contains(*cust.Packages, pkg) {
-			lintErrors = append(lintErrors, BlueprintLintItem{
+	for _, pkg := range policyBP.GetPackagesEx(false) {
+		if currentCust.Packages == nil || !slices.Contains(*currentCust.Packages, pkg) {
+			errs = append(errs, BlueprintLintItem{
 				Name:        "Compliance",
 				Description: fmt.Sprintf("package %s required by policy is not present", pkg),
 			})
 			if fixup {
-				cust.Packages = common.ToPtr(append(common.FromPtr(cust.Packages), pkg))
+				currentCust.Packages = common.ToPtr(append(common.FromPtr(currentCust.Packages), pkg))
 			}
 		}
 	}
 
-	// some policies (ansi minimal) only require some extra packages
-	if bp.Customizations == nil {
-		return lintErrors, nil
+	// Warnings: packages from saved policy no longer required
+	if prevCust != nil && prevCust.Packages != nil {
```

Comment:
I think this can be simplified further
```
if prevCust != nil {
for _, pkg := range common.FromPtr(prevCust.Packages) {
 …
}
}
```

`common.FromPtr` will return the default value in case the pointer is nil. So if prevCust.Packages is nil, it'll return an empty slice. I think quite a few things below here can be cimplified that way.